### PR TITLE
Feat/pagination in cms

### PR DESCRIPTION
**Pull request by logan-anderson**
fixes https://github.com/tinacms/product-roadmap/issues/97

### CODE REVIEW

**Comment by spbyrne** [comment]
Screen Shot 2022-05-30 at 12 22 00 PM

I've adapted the component used in the media library for styling.

**Comment by logan-anderson** [comment]
Noticing some strange pagination behaviour but not sure why its happening.

Any idea @kldavis4 ?

https://www.loom.com/share/350fae9d2b714e898b8f1d0e22339b62

**Comment by logan-anderson** [comment]
> Noticing some strange pagination behaviour but not sure why its happening.
> 
> Any idea @kldavis4 ?
> 
> https://www.loom.com/share/350fae9d2b714e898b8f1d0e22339b62

UPDATE: I have found the issue and fixed it.

File: packages/tinacms/src/admin/api.ts [hunk 1]
```diff
@@ -58,20 +58,30 @@ export class TinaAdminApi {
...
+    includeDocuments: boolean,
+    after?: string
+  ) {
     if (includeDocuments === true) {
       if (this.useDataLayer) {
         const sort = this.schema.getIsTitleFieldName(collectionName)
         const response: { collection: Collection } = await this.api.request(
           `#graphql
-      query($collection: String!, $includeDocuments: Boolean!, $sort: String){
+      query($collection: String!, $includeDocuments: Boolean!, $sort: String,  $limit: Float $after: String){
```
Comment (kldavis4) [inline packages/tinacms/src/admin/api.ts:71]: is this missing a comma?

File: packages/tinacms/src/admin/components/GetCollection.tsx [hunk 2]
```diff
@@ -66,17 +68,20 @@ const GetCollection = ({
   cms,
   collectionName,
   includeDocuments = true,
+  startCurser,
```
Comment (kldavis4) [inline packages/tinacms/src/admin/components/GetCollection.tsx:71]:
```suggestion
  startCursor,
```

File: packages/tinacms/src/admin/components/GetCollection.tsx [hunk 3]
```diff
@@ -66,17 +68,20 @@ const GetCollection = ({
...
   children,
 }: {
   cms: TinaCMS
   collectionName: string
   includeDocuments?: boolean
+  startCurser?: string
```
Comment (kldavis4) [inline packages/tinacms/src/admin/components/GetCollection.tsx:77]:
```suggestion
  startCursor?: string
```

File: packages/tinacms/src/admin/components/GetCollection.tsx [hunk 4]
```diff
@@ -66,17 +68,20 @@ const GetCollection = ({
...
   children: any
 }) => {
   const { collection, loading, error, reFetchCollection } = useGetCollection(
     cms,
     collectionName,
-    includeDocuments
+    includeDocuments,
+    startCurser || ''
```
Comment (kldavis4) [inline packages/tinacms/src/admin/components/GetCollection.tsx:84]:
```suggestion
    startCursor || ''
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 5]
```diff
@@ -120,6 +122,8 @@ const CollectionListPage = () => {
     collection: collectionName,
     relativePath: '',
   })
+  const [endCursor, setEndCurser] = useState('')
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:125]:
```suggestion
  const [endCursor, setEndCursor] = useState('')
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 6]
```diff
@@ -120,6 +122,8 @@ const CollectionListPage = () => {
...
+  const [prevCursers, setPrevCursers] = useState([])
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:126]:
```suggestion
  const [prevCursors, setPrevCursors] = useState([])
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 7]
```diff
@@ -129,11 +133,13 @@ const CollectionListPage = () => {
             cms={cms}
             collectionName={collectionName}
             includeDocuments
+            startCurser={endCursor}
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:136]:
```suggestion
            startCursor={endCursor}
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 8]
```diff
@@ -298,6 +304,26 @@ const CollectionListPage = () => {
                             </tbody>
                           </table>
                         )}
+                        <div className="pt-3">
+                          <CursorPaginator
+                            variant="white"
+                            hasNext={pageInfo.hasNextPage}
+                            navigateNext={() => {
+                              const newState = [...prevCursers, endCursor]
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:312]:
```suggestion
                              const newState = [...prevCursors, endCursor]
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 9]
```diff
@@ -298,6 +304,26 @@ const CollectionListPage = () => {
...
+                              setPrevCursers(newState)
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:313]:
```suggestion
                              setPrevCursors(newState)
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 10]
```diff
@@ -298,6 +304,26 @@ const CollectionListPage = () => {
...
+                              setEndCurser(pageInfo.endCursor)
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:314]:
```suggestion
                              setEndCursor(pageInfo.endCursor)
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 11]
```diff
@@ -298,6 +304,26 @@ const CollectionListPage = () => {
...
+                            }}
+                            hasPrev={prevCursers.length > 0}
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:316]:
```suggestion
                            hasPrev={prevCursors.length > 0}
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 12]
```diff
@@ -298,6 +304,26 @@ const CollectionListPage = () => {
...
+                            navigatePrev={() => {
+                              const prev = prevCursers[prevCursers.length - 1]
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:318]:
```suggestion
                              const prev = prevCursors[prevCursors.length - 1]
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 13]
```diff
@@ -298,6 +304,26 @@ const CollectionListPage = () => {
...
+                              if (typeof prev === 'string') {
+                                const newState = prevCursers.slice(0, -1)
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:320]:
```suggestion
                                const newState = prevCursors.slice(0, -1)
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 14]
```diff
@@ -298,6 +304,26 @@ const CollectionListPage = () => {
...
+                                setPrevCursers(newState)
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:321]:
```suggestion
                                setPrevCursors(newState)
```

File: packages/tinacms/src/admin/pages/CollectionListPage.tsx [hunk 15]
```diff
@@ -298,6 +304,26 @@ const CollectionListPage = () => {
...
+                                setEndCurser(prev)
```
Comment (kldavis4) [inline packages/tinacms/src/admin/pages/CollectionListPage.tsx:322]:
```suggestion
                                setEndCursor(prev)
```

File: packages/tinacms/src/admin/api.ts (hunk 1 above)
Comment (logan-anderson) [reply packages/tinacms/src/admin/api.ts]: yes it is. Thanks

**Comment by kldavis4** [comment]
@logan-anderson tested this on the tina-cloud-starter and it worked fine. Tried it on examples/basic without the data layer and I get this error:

Screen Shot 2022-05-31 at 11 13 01

**Review by Enigmatical — APPROVED** [review]
Pretty much exactly what I would've done 👍 Glad to see that `TinaAdminApi` paying off :smile:

Also, I like that we're reusing the `pagination` component; nice one, @spbyrne 🎉

**Pull request merged**